odoo/orm/fields_properties.py

```diff
@@ -995,6 +995,81 @@ class PropertiesDefinition(Field):
+        :param definition: List of property definitions (already processed)
+        :param separator_condition_fn: Function that takes props_in_group and returns
+                                    (should_modify_separator, new_hidden_value)
+        :return: Updated definition with processed separators
+        """
+        final_def = []
+        current_sep = None
+        current_group = []
+
+        def flush_group():
+            if not current_sep:
+                return
+
+            props_in_group = [p for p in current_group if p.get("type") != "separator"]
+            should_modify, new_hidden_value = separator_condition_fn(props_in_group)
+            if should_modify:
+                sep_copy = current_sep.copy()
+                if new_hidden_value:
+                    sep_copy["hidden"] = True
+                else:
+                    sep_copy.pop("hidden", None)
+                final_def.append(sep_copy)
+            else:
+                final_def.append(current_sep)
+            final_def.extend(current_group)
+
+        for prop in definition:
+            if prop.get("type") == "separator":
+                flush_group()
+                current_sep = prop
+                current_group = []
+            else:
+                current_group.append(prop)
+        flush_group()
+        return final_def
+
+    def set_properties_visibility(self, definition, property_names, hidden=True):
+        """
+        Set visibility of properties in the definition.
+        :param definition: List of property definitions
+        :param property_names: List of property names to modify (or single string)
+        :param hidden: True to hide properties, False to show them
+        :return: Updated definition with modified property visibility
+        """
+        if not definition:
+            return definition
+        if isinstance(property_names, str):
+            property_names = [property_names]
+        property_names_set = set(property_names)
+        new_def = []
+        for prop in definition:
+            prop_copy = prop.copy()
+            if prop_copy.get('name') in property_names_set:
+                if hidden:
+                    prop_copy['hidden'] = True
+                else:
+                    prop_copy.pop('hidden', None)
+            new_def.append(prop_copy)
+
+        def separator_condition(props_in_group):
+            if hidden:
+                # Hide separator if all properties in group are hidden
+                if props_in_group and all(p.get("hidden") for p in props_in_group):
+                    return (True, True)
+                return (False, None)
+            # Show separator if any property in group is visible
+            if any(not p.get("hidden") for p in props_in_group):
+                return (True, False)
+            return (False, None)
+
+        return self._process_property_groups(new_def, separator_condition)
```

Comment:
Since `_process_property_groups` is used once, you can inline it here.
And possibly all could be done in just one loop.